File: better/Models/Conversation.swift
import Foundation

struct Conversation: Codable, Identifiable, Equatable, Hashable {
    var id: String
    var title: String
    var createdAt: Date
    var updatedAt: Date
    var isPinned: Bool
    var isArchived: Bool
    var systemInstruction: String?
    var modelName: String

    // Generation parameters
    var temperature: Double
    var topP: Double
    var topK: Int
    var maxOutputTokens: Int

    // Media generation toggles
    var imageGenerationEnabled: Bool
    var videoGenerationEnabled: Bool
    
    // Models that have been removed — remap to default
    private static let deprecatedModels: Set<String> = [
        "deepseek/deepseek-chat",
        "moonshotai/kimi-k2.5",
        "meta-llama/llama-4-maverick:free",
        "gemini-flash-latest",
        "gemini-pro-latest",
        "gemini-2.5-flash-image",
        "gemini-3-pro-image-preview",
        "veo-3.1-generate-preview"
    ]

    init(
        id: String = UUID().uuidString,
        title: String = "New Chat",
        modelName: String = Constants.Models.defaultModel,
        temperature: Double = 1.0,
        topP: Double = 0.95,
        topK: Int = 40,
        maxOutputTokens: Int = 8192
    ) {
        self.id = id
        self.title = title
        self.createdAt = Date()
        self.updatedAt = Date()
        self.isPinned = false
        self.isArchived = false
        self.modelName = modelName
        self.temperature = temperature
        self.topP = topP
        self.topK = topK
        self.maxOutputTokens = maxOutputTokens
        self.imageGenerationEnabled = false
        self.videoGenerationEnabled = false
    }
    
    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decode(String.self, forKey: .id)
        title = try container.decode(String.self, forKey: .title)
        createdAt = try container.decode(Date.self, forKey: .createdAt)
        updatedAt = try container.decode(Date.self, forKey: .updatedAt)
        isPinned = try container.decodeIfPresent(Bool.self, forKey: .isPinned) ?? false
        isArchived = try container.decodeIfPresent(Bool.self, forKey: .isArchived) ?? false
        systemInstruction = try container.decodeIfPresent(String.self, forKey: .systemInstruction)
        let raw = try container.decode(String.self, forKey: .modelName)
        modelName = Self.deprecatedModels.contains(raw) ? Constants.Models.defaultModel : raw
        temperature = try container.decodeIfPresent(Double.self, forKey: .temperature) ?? 1.0
        topP = try container.decodeIfPresent(Double.self, forKey: .topP) ?? 0.95
        topK = try container.decodeIfPresent(Int.self, forKey: .topK) ?? 40
        maxOutputTokens = try container.decodeIfPresent(Int.self, forKey: .maxOutputTokens) ?? 8192
        imageGenerationEnabled = try container.decodeIfPresent(Bool.self, forKey: .imageGenerationEnabled) ?? false
        videoGenerationEnabled = try container.decodeIfPresent(Bool.self, forKey: .videoGenerationEnabled) ?? false
    }
}

File: web/src/models/Conversation.ts
export interface Conversation {
  id: string;
  title: string;
  createdAt: Date;
  updatedAt: Date;
  isPinned: boolean;
  isArchived: boolean;
  systemInstruction: string | null;
  modelName: string;
  temperature: number;
  topP: number;
  topK: number;
  maxOutputTokens: number;
  thinkingBudget: number | null;
  googleSearchEnabled: boolean;
  codeExecutionEnabled: boolean;
  urlContextEnabled: boolean;
  imageGenerationEnabled: boolean;
  videoGenerationEnabled: boolean;
}

export type NewConversationInput = Omit<Conversation, 'id' | 'createdAt' | 'updatedAt'>;

const DEFAULT_TITLE = 'New Chat';
const DEFAULT_MODEL_NAME = 'gemini-flash-latest';

const generateId = (): string => {
  if (typeof crypto !== 'undefined' && typeof crypto.randomUUID === 'function') {
    return crypto.randomUUID();
  }

  return `conv_${Date.now()}_${Math.random().toString(16).slice(2)}`;
};

export const createConversation = (title: string = DEFAULT_TITLE): Conversation => ({
  id: generateId(),
  title,
  createdAt: new Date(),
  updatedAt: new Date(),
  isPinned: false,
  isArchived: false,
  systemInstruction: null,
  modelName: DEFAULT_MODEL_NAME,
  temperature: 1.0,
  topP: 0.95,
  topK: 40,
  maxOutputTokens: 8192,
  thinkingBudget: null,
  googleSearchEnabled: false,
  codeExecutionEnabled: false,
  urlContextEnabled: false,
  imageGenerationEnabled: false,
  videoGenerationEnabled: false,
});

const THINKING_MODEL_PATTERNS: RegExp[] = [/pro/i, /thinking/i];

export const isThinkingModel = (modelName: string): boolean =>
  THINKING_MODEL_PATTERNS.some((pattern) => pattern.test(modelName));

export const getDefaultConversationSettings = (): Partial<Conversation> => {
  return {
    modelName: DEFAULT_MODEL_NAME,
    temperature: 1.0,
    topP: 0.95,
    topK: 40,
    maxOutputTokens: 8192,
    thinkingBudget: null,
    googleSearchEnabled: false,
    codeExecutionEnabled: false,
    urlContextEnabled: false,
    imageGenerationEnabled: false,
    videoGenerationEnabled: false,
    systemInstruction: null
  };
};
